### ISSUE TITLE
Improve coverage collection performance for large projects

### ISSUE THREAD

**Issue by orestesgaolin** [flutter/flutter#80516]
## Steps to Reproduce

1. Run all tests in [following repository](https://github.com/orestesgaolin/coverage_test_regression)
    ```
    flutter test --no-pub --coverage --test-randomize-ordering-seed random -j 28 --no-track-widget-creation
    flutter test --no-pub --test-randomize-ordering-seed random -j 28 --no-track-widget-creation
    ```
2. Observe the difference in time elapsed

**Expected results:** 

Similar test duration in both cases

**Actual results:** 

Tests with coverage take significantly more time, especially in CI environment where you can't take advantage of multi core execution.

For instance on my computer the first run _with coverage_:

```
flutter clean
flutter pub get
flutter test --no-pub --coverage --test-randomize-ordering-seed random -j 28 --no-track-widget-creation
...
00:14 +8: All tests passed!
```

and immediately after _without_:

```
flutter test --no-pub --test-randomize-ordering-seed random -j 28 --no-track-widget-creation
...
00:04 +8: All tests passed!
```

Calling it again _with coverage_ results with very short run (about 0:04) but after `flutter clean` the following call _without coverage_ gives relatively shorter run:

```
flutter clean
flutter pub get
flutter test --no-pub --test-randomize-ordering-seed random -j 28 --no-track-widget-creation
...
00:08 +8: All tests passed!
```

See [this workflow run](https://github.com/orestesgaolin/coverage_test_regression/pull/2/checks?check_run_id=2353871491) where the difference is even larger.

![image](https://user-images.githubusercontent.com/16854239/114889725-a466e380-9e0a-11eb-9f3b-b7a8ac59d797.png)

Sample app https://github.com/orestesgaolin/coverage_test_regression

In our production app where there's almost 1700 tests (many widget tests) the difference is even more striking (16-18 minutes vs 2-3 minutes).

flutter doctor

```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 2.0.4, on macOS 11.2.3 20D91 darwin-x64, locale pl-PL)
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
[✓] Xcode - develop for iOS and macOS
[✓] Chrome - develop for the web
[✓] Android Studio (version 4.1)
[✓] VS Code (version 1.55.1)
[✓] Connected device (4 available)

• No issues found!
```

This is a follow up from #23157
cc: @jonahwilliams

**Comment by jonahwilliams** [comment]
> Calling it again with coverage results with very short run (about 0:04) but after flutter clean the following call without coverage gives relatively shorter run:

FYI this is expected since tests, with or without coverage, cache the compiled code for faster invalidation.

**Comment by jonahwilliams** [comment]
I cannot version solve your example app on ToT:

```
Because no versions of bloc_test match >8.0.0 <9.0.0 and bloc_test 8.0.0 depends on test ^1.16.0, bloc_test ^8.0.0 requires test ^1.16.0.
And because test >=1.16.6 depends on coverage ^1.0.1 and test >=1.16.0-nullsafety.19 <1.16.6 depends on test_api 0.2.19, bloc_test ^8.0.0 requires coverage ^1.0.1 or test_api 0.2.19.
And because coverage depends on flutter_test any from sdk which depends on test_api 0.3.0, bloc_test ^8.0.0 requires coverage ^1.0.1.
So, because coverage depends on bloc_test ^8.0.0 and coverage is 1.0.0+1, version solving failed.
```

**Comment by orestesgaolin** [comment]
> I cannot version solve your example app on ToT:

looking into that

**Comment by jonahwilliams** [comment]
I rolled back to stable in order to verify the performance but it might be worth keeping in mind

**Comment by jonahwilliams** [comment]
On the current 2.0 stable:

```
PS C:\Users\Jonah\Documents\coverage_test_regression> flutter test
00:03 +8: All tests passed!
PS C:\Users\Jonah\Documents\coverage_test_regression> flutter test --coverage
00:02 +8: All tests passed!
```

Performance seems to be pretty close

**Comment by jonahwilliams** [comment]
Is the performance only worse after a flutter clean? if so that is entirely expected due to https://github.com/flutter/flutter/issues/80516#issuecomment-820549293

**Comment by orestesgaolin** [comment]
I updated the name of the app from `coverage` to `coverageapp`, it wasn't the smartest choice of app name, pub get should work now on master ;)


On current master 2.2.0-11.0.pre.127

```sh
# coverage
❯ f test --no-pub --coverage --test-randomize-ordering-seed random -j 28 --no-track-widget-creation
Shuffling test order with --test-randomize-ordering-seed=709197374
00:00 +0: loading /Users/dominik/Projects/_vgv/coverage/test/app_test.dart                                                                  
 ...
00:14 +0: loading /Users/dominik/Projects/_vgv/coverage/test/app_test.dart                                                                      
00:14 +0: /Users/dominik/Projects/_vgv/coverage/test/counter/cubit/counter_cubit_test.dart: CounterCubit initial state is 0
00:16 +8: All tests passed!

# no coverage
❯ f test --no-pub --test-randomize-ordering-seed random -j 28 --no-track-widget-creation
...
00:05 +8: All tests passed!

❯ f clean
Cleaning Xcode workspace...                                         6,0s
Deleting build...                                                    2ms
Deleting .dart_tool...                                               0ms
Deleting .packages...                                                0ms
Deleting Generated.xcconfig...                                       0ms
Deleting flutter_export_environment.sh...                            0ms

❯ f pub get
Running "flutter pub get" in coverage...                         1 948ms

# no coverage
❯ f test --no-pub --test-randomize-ordering-seed random -j 28 --no-track-widget-creation
Shuffling test order with --test-randomize-ordering-seed=3980231199
00:00 +0: loading /Users/dominik/Projects/_vgv/coverage/test/app_test.dart                                                                      
 ...
00:09 +8: All tests passed!

```

_f is for flutter_

The main issue is that the first run, especially on CI, is really slow. The `loading` step takes vast majority of the time (00:14 out of 00:16) in this case but in our production app it's much less visible. When coverage is enabled it's 3 times slower than subsequent run without coverage (00:16 vs 00:05) and almost 2 times slower than first run without coverage (00:09)


Here's a sample log from our CI from couple days ago [10_Run tests.txt](https://github.com/flutter/flutter/files/6319680/10_Run.tests.txt), when the test took 17 minutes to run and the loading step didn't take more than couple of seconds. It's very hard for me to distil exact cause of this long duration, but I assume it's a combination of number of tests, their complexity and overhead of coverage capture. In my simple example above the coverage tests take twice as much time as normal, but in real production app it's several times more :/ I wonder if there's any way to cut down this time, because in our case even 10-15% less would get significant savings.

**Comment by jonahwilliams** [comment]
If the main issue is the first run - that doesn't really seem to be related to coverage. The first run performance is likely getting blocked by the initial compilation

**Comment by orestesgaolin** [comment]
> The first run performance is likely getting blocked by the initial compilation

Would there be any chance to improve this if that's the case?

I just modified the steps and run our 1600+ tests on GitHub Actions and:

- without coverage: 04:12
- with coverage immediately after: 19:38

![image](https://user-images.githubusercontent.com/16854239/114916426-10eedc00-9e25-11eb-8e8d-a43d328aee6b.png)


So there's almost no benefit of running the tests _without coverage_ before tests _with coverage_. Wouldn't the compilation happen during the first run (_without coverage_) so that the second run (_with coverage_) doesn't need to compile it again?

Is there any way to profile the tests to see if there's any bottlenecks? Maybe some kind of pattern that we use in tests or library?

**Comment by jonahwilliams** [comment]
Ahh OK thank you for the additional information there. Its unfortunate that this performance difference doesn't seem to reproduce locally. A coverage test is going to be a lot more "chatty" over loopback, and I wonder if that is an issue that is specific to that CI setup.

**Comment by jonahwilliams** [comment]
I could create a branch that users fewer network requests, @orestesgaolin if you'd be willing to run that through your CI to see the performance difference that would help me a lot

**Comment by orestesgaolin** [comment]
I'm happy to test if you can prepare something. 

Right now I'm trying to investigate what has the largest impact on tests: widget or unit tests and this is the breakdown. Seems like removing around 830 widget tests cuts around 13 minutes from the test run which is much more than half of the original time.

| Case | Unit + widget (1685) | Unit only (849) |
| --- | --- | --- |
| No coverage | 04:12 | 01:12 |
| Coverage | 19:38 | 06:44 |

**Comment by jonahwilliams** [comment]
For a quick example of caching some of the network chatter, try https://github.com/flutter/flutter/pull/80526

**Comment by jonahwilliams** [comment]
So some investigation on my end: there isn't a regression I can pinpoint - instead it seems like as the number of libraries under test grow, the coverage measurement slows down. I'll see if I can revisit some of the more aggressive optimizations

**Comment by jonahwilliams** [comment]
I did so more research and I don't have any way to improve the performance without changing the VM Service API we use to collect coverage. unfortunately as the number of libraries grows test coverage collection will slow down.

**Comment by jonahwilliams** [comment]
Here is a timeline trace of a coverage collection of the flutter framework. With a breakpoint set before/after coverage collection in the tool:

` dart --observe ..\flutter_tools\bin\flutter_tools.dart test --coverage .\test\material\animated_icons_test.dart`


https://gist.github.com/jonahwilliams/0fa8f9e945814087c39b75139218b282

**Comment by jonahwilliams** [comment]
This also transfers ~64 MB of data in total, which I measured by appending all received JSON into a single file.

**Comment by aniketudeshmukh** [comment]
Any update on this issue?

**Comment by byronkimani** [comment]
Any updates??

**Comment by vincevargadev** [comment]
Not a solution to the problem, but adding a higher custom timeout to the tests with coverage, for example `--timeout 120s`, solved our issues on our CI/CD pipeline.

**Comment by orestesgaolin** [comment]
Another issue popped up with similar problem https://github.com/flutter/flutter/issues/86722

**Comment by orestesgaolin** [comment]
Related issue https://github.com/dart-lang/sdk/issues/45985

**Comment by a-siva** [comment]
I am going to mark this issue as a duplicate of https://github.com/flutter/flutter/issues/86722 as most of the work and discussion is happening in that issue.
